### ISSUE TITLE
Questions regarding replication results of CCT5 on the JIT-DP task.

### ISSUE THREAD

**Issue by Hugo-Liang** [Ringbo/CCT5#4]
Hi, 

I tried to replicate the experimental results of CCT5 on the JIT-DP task, but found some problems during resolving the encountered issues.

1. Since the Hugging Face site is not allowed to access on my cloud server, I modified the 'model_name_or_path' (in scripts/finetune_jitdp_SF.sh) parameter to local codet5_base folder downloaded from Hugging Face. e.g.
`--model_name_or_path ${CURRENT_DIR}/models/codet5_base \`
Similarily, the tokenizer path (in src/models.py) is modified to local roberta_base folder downloaded from Hugging Face. e.g.
```
roberta_tokenizer_path = os.path.abspath(os.path.join(current_dir, '..', 'models/roberta_base/'))
tokenizer = tokenizer_class.from_pretrained(roberta_tokenizer_path)
```

However, on a 3090GPU, the result of F1 and AUC is 0.395 and 0.871, respectively. The replicated F1 score is less than the 0.451 reported in the paper.

2. In the defined build_or_load_gen_model function (src/models.py), it seems the codet5_base mode is loaded (via the 'model_name_or_path' parameter) firstly, then enrich_vocab and resize_token_embeddings for the loaded model, the pretrained state dict of CCT5 is loaded (via the 'load_model_path' parameter) at last. So I guess the pretrained CCT5 should be loaded first , or invalid the step of loading pretrained state dict of CCT5. e.g. In scripts/finetune_jitdp_SF.sh,
using `--model_name_or_path $PRETRAINED_MODEL_DIR\`
or delete `--load_model_path $MODEL_PATH \`

After that, the result of F1 and AUC is 0.435 and 0.871, respectively. The F1 score is still less than 0.451.

3. I also tried the above steps on a P40 GPU, while the results are more less than those got on 3090.

I may misunderstood something or did something wrong. Could you help to point out my mistakes or give some comments?

Thank you,
Hugo

**Comment by Ringbo** [comment]
Hi Hugo,

Thanks for your intersting on our work. The issues you've encountered might be due to the different execution environments; for instance, different version of Pytorch might slightly. This may influence the training result especially in tasks with small dataset, such as JIT-DP.
I once experienced a similar situation where the training results were different due to different CUDA versions, although it wasn't in this project. I recommend using the same version of libraries as in CCT5 or reusing our released checkpoint for the JIT-DP task  to avoid the potential bias. I've tested it multiple times, and the released checkpoint shows consistent performance on different machines.

Best,
Bo.

**Comment by Hugo-Liang** [comment]
Hi Bo,


Thanks for your quick reply and apologize for my late response. I was able to access the 3090 until yesterday. Before that, I did some tests on a P40.

1. I cloned this project to a new folder again, following the scripts (described in **src/models.py** and **scripts/finetune_jitdp_SF.sh**) strictly. After that, the results of F1 is 0.4277, while this number will be 0.4136 if I modified the `--model_name_or_path` parameter from `${CURRENT_DIR}/models/codet5_base` to` $PRETRAINED_MODEL_DIR` or invalid the setting of `--load_model_path $MODEL_PATH`. Therefore, in the last reply, my understanding for the process of loading CCT5 should be incorrect.

2. I carefully checked versions of packages installed in the conda environment, only the tree-sitter is different of that described in the **scripts/setup.sh**. I installed the version of 0.19.0 instead of 0.19.1 because of the PyPi mirror. Since this dependency is not used during using the downloaded files and checkpoints for finetuning pretrained CCT5 on the JIT-DP task, I think the effect may be ignored.

3. I also tried to use different CUDA versions, for example, torch=2.0.0+cu118 rather than the default torch=2.0.0+cu117, and get exact same results. As you said, this issue "wasn't in this project". Besides, the result of F1 and AUC described in paper can be replicated by using the released fine-tuned checkpoint (`FINETUNED_MODEL_PATH="$CURRENT_DIR/models/fine-tuning/JITDefectPrediction/SF/pytorch_model.bin"`).

4. In my side, the results obtained from P40 is always less than 3090, and keep consistent if the GPU type is fixed. So the ideal result of F1 could fall into the range of [0.44, 0.45] in my guess. Therefore, the F1 result of 0.395 on 3090 is really odd, and force me to scrutinize the code and do more tests.

5. Currently, the results of F1 and AUC remain 0.395 and 0.871 on 3090, whatever using cu117 or cu118. I noticed the `--train_steps` is set to 2500 in **scripts/finetune_jitdp_SF_EF.sh**, so I modified this parameter in **finetune_jitdp_SF.sh**, but found the hope of "A better checkpoint may be found with more training steps" in vain. Actually, the F1 result became even worth (0.2635). It seems more exploration are needed to sort out what has happened. I would attribute this phenomenon as a bug in 3090 at present, and will do related evaluation or comparison experiments on P40 in later.

7. Some screenshots of the saved checkpoints are attached, in case you or someone are interested. Any insights would be appreciated and thanks again for your time and efforts. This work is inspiring!


Best Regards,
Hugo
<img width="220" alt="3" src="https://github.com/Ringbo/CCT5/assets/103502294/84681407-9138-4921-86f9-f879c39bb99d">
<img width="220" alt="2" src="https://github.com/Ringbo/CCT5/assets/103502294/8d435da0-d473-4b37-bdfe-2a9b5ea64d41">
<img width="228" alt="1" src="https://github.com/Ringbo/CCT5/assets/103502294/310d9611-944b-4609-9e7b-7576022cf01b">
